aaelf32/aaelf32.rst

```diff
@@ -0,0 +1,3289 @@
+
+.. rubric:: Group relocations
+
+Relocation codes 4 and 57-83 are intended to relocate sequences of instructions
+that generate a single address.  They are encoded to extract the maximum
+flexibility from the Arm ADD- and SUB-immediate instructions without need to
+determine during linking the full sequence being used.  The relocations operate
+by performing the basic relocation calculation and then partitioning the result
+into a set of groups of bits that can be statically determined.  All processing
+for the formation of the groups is done on the absolute value of X; the sign of
+X is used to determine whether ADD or SUB instructions are used, or, if the
+sequence concludes with a load/store operation, the setting of the U bit (bit
+23) in the instruction.
+
+A group, G\ :sub:`n`, is formed by examining the residual value, Y\ :sub:`n`, after
+the bits for group G\ :sub:`n–1` have been masked off.  Processing for group
+G\ :sub:`0` starts with the absolute value of X.  For ALU-type relocations a
+group is formed by determining the most significant bit (MSB) in the residual
+and selecting the smallest constant K\ :sub:`n` such that
+
+  MSB(Y\ :sub:`n`) & (255 << 2K\ :sub:`n`) != 0,
+
+except that if Y\ :sub:`n` is 0, then K\ :sub:`n` is 0.  The value G\ :sub:`n` is
+then
+
+  Y\ :sub:`n` & (255 << 2K\ :sub:`n`),
+
+and the residual, Y\ :sub:`n+1`, for the next group is
+
+  Y\ :sub:`n` & ~G\ :sub:`n`.
+
+Note that if Y\ :sub:`n` is 0, then G\ :sub:`n` will also be 0.
+
+For group relocations that access memory the residual value is examined in its
+entirety (i.e. after the appropriate sequence of ALU groups have been removed):
+if the relocation has not overflowed, then the residual for such an instruction
+will always be a valid offset for the indicated type of memory access.
+
+Overflow checking is always performed on the highest-numbered group in a
+sequence.  For ALU-type relocations the result has overflowed if Y\ :sub:`n+1` is not
+zero.  For memory access relocations the result has overflowed if the residual
+is not a valid offset for the type of memory access.
+
+.. note::
+   The unchecked (_NC) group relocations all include processing of the Thumb
+   bit of a symbol.  However, the memory forms of group relocations (eg
+   R_ARM_LDR_G0) ignore this bit.  Therefore the use of the memory forms with
+   symbols of type STT_FUNC is unpredictable.
+
+Static Thumb16 relocations
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Relocations for 16-bit thumb instructions are shown in the following table. In
+general the addressing range of these relocations is too small for them to
+reference external symbols and they are documented here for completeness. A
+linker is not required to generate trampoline sequences (or veneers) to extend
+the branching range of the jump relocations.
+
+Relocation R_ARM_THM_JUMP6 is only applicable to the Thumb-2 instruction set.
+
+.. _Static Thumb-16 relocations:
+
+.. table:: Static Thumb-16 relocations
+  :class: longtable
```

Comment:
as described above on the version history table, please use a class to specify the width, and no longtable